linera-service/src/proxy/grpc.rs

```diff
@@ -468,6 +470,44 @@ where
         Ok(Response::new(certificate.try_into()?))
     }
 
+    #[instrument(skip_all, err(Display))]
+    async fn download_certificates(
+        &self,
+        request: Request<CertificatesBatchRequest>,
+    ) -> Result<Response<CertificatesBatchResponse>, Status> {
+        let hashes = request
+            .into_inner()
+            .hashes
+            .into_iter()
+            .map(linera_base::crypto::CryptoHash::try_from)
+            .collect::<Result<Vec<linera_base::crypto::CryptoHash>, _>>()?;
+        let mut certificates: Vec<linera_chain::data_types::Certificate> = self
+            .0
+            .storage
+            .read_certificates(hashes)
+            .await
+            .map_err(|err| Status::from_error(Box::new(err)))?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        // Use 70% of the max message size as a buffer capacity.
+        // Leave 30% as overhead.
+        let message_capacity = (GRPC_MAX_MESSAGE_SIZE as f64 * 0.7) as usize;
+
+        // Keep dropping elements until we fit in the limited message size.
+        while !try_proto_serialize::<
```

Comment:
If a malicious client requested waaay too many certificates, this would do O((waaay too many)²) serialization work.

I'd rather serialize all the certificates individually once, and stop before we cross the 70%. The overhead capacity should then be enough to allow serializing the vector itself, too.

Reply:
Ofc, you're right. I even woke up this morning with a plan to do that.

Reply:
Done in [2cd7591](https://github.com/linera-io/linera-protocol/pull/2644/commits/2cd759129623bc35a9fddd75e389952fe0cf52b5). PTAL

Reply:
I changed the approach since I couldn't get the `&mut [u8]` and `impl BufMut` (from `protoc::Message`) to work as expected (ie it didn't advance the pointer to the underlying slice and I could fit 2 elements into a slice for a vec that should have fit only 1 element). I decided to go a step further and just track the remaining byte limits that manually.